funsor/optimizer.py

```diff
@@ -197,17 +205,14 @@ def optimize_reduction(op, arg, reduced_vars):
         # count new appearance of variables that aren't reduced
         reduce_dim_counter.update({d: 1 for d in reduced_vars & (both_vars - path_end_reduced_vars)})
 
-        path_end = Binary(finitary_op, ta, tb)
-        if path_end_reduced_vars:
-            path_end = Reduce(reduce_op, path_end, path_end_reduced_vars)
-
+        path_end = Contract(sum_op, prod_op, ta, tb, path_end_reduced_vars)
```

Comment:
Note that the path optimizer can be called recursively here if `ta` or `tb` are `Finitary`